readthedocs/restapi/views/model_views.py

```diff
@@ -92,11 +92,16 @@ class ProjectViewSet(UserSelectViewSet):
     model = Project
     pagination_class = api_utils.ProjectPagination
 
+    def get_project_for_user_or_404(self, lookup_value):
+        lookup_query = {self.lookup_field: lookup_value}
+        qs = self.get_queryset()
+        return get_object_or_404(qs, **lookup_query)
+
     @decorators.detail_route()
     def valid_versions(self, request, **kwargs):
         """Maintain state of versions that are wanted."""
-        project = get_object_or_404(
-            Project.objects.api(request.user), pk=kwargs['pk'])
```

Comment:
`UserSelectViewSet.get_queryset` looks very much the same code

Reply:
`Project.objects.api` is overriden in our commercial site to handle different kind of permissions. That's the problem here.

This is the code used to extend this class from _outside_:

https://github.com/rtfd/readthedocs.org/blob/master/readthedocs/projects/querysets.py#L66-L69

Even if `get_queryset` is too similar here, it will break our commercial site and we will need to refactor that code also.

I hope that helps you to understand why we can't accept this change for now.

Reply:
What am trying to tell you is that `UserSelectViewSet.get_queryset` is a fancy way to write `return Project.objects.api(self.request.user)`.
https://github.com/rtfd/readthedocs.org/blob/master/readthedocs/restapi/views/model_views.py#L79